common/storage/sql/sqlite3/storage.c

```diff
@@ -33,102 +32,167 @@ retcode_t iota_stor_destroy() {
+  tx->timestamp = sqlite3_column_int64(statement, 4);
+  tx->current_index = sqlite3_column_int64(statement, 5);
+  tx->last_index = sqlite3_column_int64(statement, 6);
+  buffer = sqlite3_column_blob(statement, 7);
+  memcpy(tx->bundle, buffer, FLEX_TRIT_SIZE_243);
+  buffer = sqlite3_column_blob(statement, 8);
+  memcpy(tx->trunk, buffer, FLEX_TRIT_SIZE_243);
+  buffer = sqlite3_column_blob(statement, 9);
+  memcpy(tx->branch, buffer, FLEX_TRIT_SIZE_243);
+  buffer = sqlite3_column_blob(statement, 10);
+  memcpy(tx->tag, buffer, FLEX_TRIT_SIZE_81);
+  tx->attachment_timestamp = sqlite3_column_int64(statement, 11);
+  tx->attachment_timestamp_upper = sqlite3_column_int64(statement, 12);
+  tx->attachment_timestamp_lower = sqlite3_column_int64(statement, 13);
+  buffer = sqlite3_column_blob(statement, 14);
+  memcpy(tx->nonce, buffer, FLEX_TRIT_SIZE_81);
+  buffer = sqlite3_column_blob(statement, 15);
+  memcpy(tx->hash, buffer, FLEX_TRIT_SIZE_243);
 }
 
 retcode_t iota_stor_store(const connection_t* const conn,
                           const iota_transaction_t data_in) {
   char statement[MAX_STORE_STATEMENT_SIZE];
   iota_transactions_insert_statement(data_in, statement,
                                      MAX_STORE_STATEMENT_SIZE);
-
   char* err_msg = 0;
-  int rc = sqlite3_exec((sqlite3*)conn->db, statement, 0, 0, &err_msg);
+  sqlite3_stmt* sqlite_statement = 0;
+  char* buffer;
+
+  int rc = sqlite3_prepare_v2((sqlite3*)conn->db, statement, -1,
+                              &sqlite_statement, &err_msg);
+  if (rc != SQLITE_OK) {
+    log_error(SQLITE3_LOGGER_ID,
+              "Failed preparing statement, sqlite3 code is: %\" PRIu64 \"", rc);
+    return RC_SQLITE3_FAILED_PREPARED_STATEMENT;
+  }
+  sqlite3_bind_blob(sqlite_statement, 1, data_in->signature_or_message,
+                    FLEX_TRIT_SIZE_6561, NULL);
+  sqlite3_bind_blob(sqlite_statement, 2, data_in->address, FLEX_TRIT_SIZE_243,
+                    NULL);
+  sqlite3_bind_blob(sqlite_statement, 3, data_in->obsolete_tag,
+                    FLEX_TRIT_SIZE_81, NULL);
+  sqlite3_bind_blob(sqlite_statement, 4, data_in->bundle, FLEX_TRIT_SIZE_243,
+                    NULL);
+  sqlite3_bind_blob(sqlite_statement, 5, data_in->trunk, FLEX_TRIT_SIZE_243,
+                    NULL);
+  sqlite3_bind_blob(sqlite_statement, 6, data_in->branch, FLEX_TRIT_SIZE_243,
+                    NULL);
+  sqlite3_bind_blob(sqlite_statement, 7, data_in->tag, FLEX_TRIT_SIZE_81, NULL);
+  sqlite3_bind_blob(sqlite_statement, 8, data_in->nonce, FLEX_TRIT_SIZE_81,
+                    NULL);
+  sqlite3_bind_blob(sqlite_statement, 9, data_in->hash, FLEX_TRIT_SIZE_243,
+                    NULL);
+  sqlite3_step(sqlite_statement);
+  if (rc != SQLITE_OK) {
+    log_error(SQLITE3_LOGGER_ID,
+              "Failed in step, sqlite3 code is: %\" PRIu64 \"", rc);
+    return RC_SQLITE3_FAILED_STEP;
+  }
 
+  rc = sqlite3_finalize(sqlite_statement);  //  Finalize the prepared statement.
```

Comment:
sqlite3 prepared statements are reusable (see `sqlite3_clear_bindings`) but are *not* thread-safe.

Can you add an issue for adding the prepared statements to the connection handle?
However, we'll still need to figure out how to guarantee thread-safety in this case (TLS?)

Reply:
sure